# Results of Multigrid for Reaction-Diffusion

## Imports

In [1]:
from main import inputParametersBPX, runDemoBPX, runDemoHB
import logging
logging.basicConfig(level=logging.DEBUG , format='[%(asctime)s] - [%(levelname)s] - %(message)s')

## Set parameters

In [2]:
inputParametersBPX()

* DEFAULT VALUES *
LEVELS              = 5
SHOW_GRIDS          = False
DEGREE              = 1
MAX_ITER            = 100
EPSILON             = 1e-08

Would you like to change some values? (Y/N): n


## Demo

Now that the user has inserted the parameters of the multigrid algorithm,
we start by solving the two-grid method, refine the grid, solve the three-grid method, etc.

In [3]:
#runDemoBPX()

In [4]:
runDemoHB()

[2020-07-17 21:49:08,095] - [DEBUG] - __init__ took 0 seconds 215 milliseconds.
[2020-07-17 21:49:08,096] - [INFO] - +-----------------------------------------+
[2020-07-17 21:49:08,097] - [INFO] - +    MULTIGRID (LEVELS =  2)              +
[2020-07-17 21:49:08,097] - [INFO] - +-----------------------------------------+
[2020-07-17 21:49:08,374] - [INFO] - Number of DoFs: 21 (by level: 8,21)
[2020-07-17 21:49:08,376] - [DEBUG] - __call__ took 0 seconds 1 milliseconds.
[2020-07-17 21:49:08,377] - [DEBUG] - __call__ took 0 seconds 1 milliseconds.
[2020-07-17 21:49:08,378] - [DEBUG] - 0.th residual: 0.4583333879709244
[2020-07-17 21:49:08,378] - [DEBUG] - __call__ took 0 seconds 0 milliseconds.
[2020-07-17 21:49:08,379] - [DEBUG] - 1.th residual: 0.4940118189204678
/home/julian/Code/Multigrid/MultigridPython/solver.py:82: RuntimeWarning: invalid value encountered in double_scalars
  alpha = r_tilde.dot(r)/d.dot(z) # alpha = (\tilde{r}^T * r) / (d^T * z)
/home/julian/Code/Multigrid/Multig

[2020-07-17 21:49:08,475] - [DEBUG] - 54.th residual: nan
[2020-07-17 21:49:08,477] - [DEBUG] - __call__ took 0 seconds 0 milliseconds.
[2020-07-17 21:49:08,478] - [DEBUG] - 55.th residual: nan
[2020-07-17 21:49:08,479] - [DEBUG] - __call__ took 0 seconds 0 milliseconds.
[2020-07-17 21:49:08,479] - [DEBUG] - 56.th residual: nan
[2020-07-17 21:49:08,480] - [DEBUG] - __call__ took 0 seconds 0 milliseconds.
[2020-07-17 21:49:08,480] - [DEBUG] - 57.th residual: nan
[2020-07-17 21:49:08,481] - [DEBUG] - __call__ took 0 seconds 1 milliseconds.
[2020-07-17 21:49:08,482] - [DEBUG] - 58.th residual: nan
[2020-07-17 21:49:08,484] - [DEBUG] - __call__ took 0 seconds 1 milliseconds.
[2020-07-17 21:49:08,485] - [DEBUG] - 59.th residual: nan
[2020-07-17 21:49:08,485] - [DEBUG] - __call__ took 0 seconds 0 milliseconds.
[2020-07-17 21:49:08,486] - [DEBUG] - 60.th residual: nan
[2020-07-17 21:49:08,487] - [DEBUG] - __call__ took 0 seconds 0 milliseconds.
[2020-07-17 21:49:08,487] - [DEBUG] - 61.th resi

[2020-07-17 21:49:08,758] - [DEBUG] - __call__ took 0 seconds 1 milliseconds.
[2020-07-17 21:49:08,758] - [DEBUG] - 9.th residual: nan
[2020-07-17 21:49:08,760] - [DEBUG] - __call__ took 0 seconds 1 milliseconds.
[2020-07-17 21:49:08,760] - [DEBUG] - 10.th residual: nan
[2020-07-17 21:49:08,761] - [DEBUG] - __call__ took 0 seconds 1 milliseconds.
[2020-07-17 21:49:08,762] - [DEBUG] - 11.th residual: nan
[2020-07-17 21:49:08,763] - [DEBUG] - __call__ took 0 seconds 1 milliseconds.
[2020-07-17 21:49:08,764] - [DEBUG] - 12.th residual: nan
[2020-07-17 21:49:08,765] - [DEBUG] - __call__ took 0 seconds 1 milliseconds.
[2020-07-17 21:49:08,766] - [DEBUG] - 13.th residual: nan
[2020-07-17 21:49:08,767] - [DEBUG] - __call__ took 0 seconds 1 milliseconds.
[2020-07-17 21:49:08,768] - [DEBUG] - 14.th residual: nan
[2020-07-17 21:49:08,769] - [DEBUG] - __call__ took 0 seconds 0 milliseconds.
[2020-07-17 21:49:08,770] - [DEBUG] - 15.th residual: nan
[2020-07-17 21:49:08,771] - [DEBUG] - __call__ to

[2020-07-17 21:49:08,872] - [DEBUG] - 69.th residual: nan
[2020-07-17 21:49:08,873] - [DEBUG] - __call__ took 0 seconds 0 milliseconds.
[2020-07-17 21:49:08,874] - [DEBUG] - 70.th residual: nan
[2020-07-17 21:49:08,875] - [DEBUG] - __call__ took 0 seconds 1 milliseconds.
[2020-07-17 21:49:08,876] - [DEBUG] - 71.th residual: nan
[2020-07-17 21:49:08,877] - [DEBUG] - __call__ took 0 seconds 0 milliseconds.
[2020-07-17 21:49:08,877] - [DEBUG] - 72.th residual: nan
[2020-07-17 21:49:08,878] - [DEBUG] - __call__ took 0 seconds 0 milliseconds.
[2020-07-17 21:49:08,879] - [DEBUG] - 73.th residual: nan
[2020-07-17 21:49:08,879] - [DEBUG] - __call__ took 0 seconds 0 milliseconds.
[2020-07-17 21:49:08,880] - [DEBUG] - 74.th residual: nan
[2020-07-17 21:49:08,881] - [DEBUG] - __call__ took 0 seconds 0 milliseconds.
[2020-07-17 21:49:08,882] - [DEBUG] - 75.th residual: nan
[2020-07-17 21:49:08,883] - [DEBUG] - __call__ took 0 seconds 0 milliseconds.
[2020-07-17 21:49:08,883] - [DEBUG] - 76.th resi

[2020-07-17 21:49:09,278] - [DEBUG] - 23.th residual: 0.25403645524719337
[2020-07-17 21:49:09,280] - [DEBUG] - __call__ took 0 seconds 1 milliseconds.
[2020-07-17 21:49:09,280] - [DEBUG] - 24.th residual: 0.25403645524719337
[2020-07-17 21:49:09,282] - [DEBUG] - __call__ took 0 seconds 1 milliseconds.
[2020-07-17 21:49:09,283] - [DEBUG] - 25.th residual: 0.25403645524719337
[2020-07-17 21:49:09,284] - [DEBUG] - __call__ took 0 seconds 1 milliseconds.
[2020-07-17 21:49:09,285] - [DEBUG] - 26.th residual: 0.25403645524719337
[2020-07-17 21:49:09,286] - [DEBUG] - __call__ took 0 seconds 2 milliseconds.
[2020-07-17 21:49:09,286] - [DEBUG] - 27.th residual: 0.25403645524719337
[2020-07-17 21:49:09,288] - [DEBUG] - __call__ took 0 seconds 1 milliseconds.
[2020-07-17 21:49:09,288] - [DEBUG] - 28.th residual: nan
[2020-07-17 21:49:09,289] - [DEBUG] - __call__ took 0 seconds 1 milliseconds.
[2020-07-17 21:49:09,290] - [DEBUG] - 29.th residual: nan
[2020-07-17 21:49:09,291] - [DEBUG] - __call__

[2020-07-17 21:49:09,401] - [DEBUG] - 83.th residual: nan
[2020-07-17 21:49:09,403] - [DEBUG] - __call__ took 0 seconds 2 milliseconds.
[2020-07-17 21:49:09,404] - [DEBUG] - 84.th residual: nan
[2020-07-17 21:49:09,405] - [DEBUG] - __call__ took 0 seconds 1 milliseconds.
[2020-07-17 21:49:09,406] - [DEBUG] - 85.th residual: nan
[2020-07-17 21:49:09,407] - [DEBUG] - __call__ took 0 seconds 1 milliseconds.
[2020-07-17 21:49:09,408] - [DEBUG] - 86.th residual: nan
[2020-07-17 21:49:09,410] - [DEBUG] - __call__ took 0 seconds 0 milliseconds.
[2020-07-17 21:49:09,410] - [DEBUG] - 87.th residual: nan
[2020-07-17 21:49:09,412] - [DEBUG] - __call__ took 0 seconds 1 milliseconds.
[2020-07-17 21:49:09,412] - [DEBUG] - 88.th residual: nan
[2020-07-17 21:49:09,414] - [DEBUG] - __call__ took 0 seconds 1 milliseconds.
[2020-07-17 21:49:09,414] - [DEBUG] - 89.th residual: nan
[2020-07-17 21:49:09,416] - [DEBUG] - __call__ took 0 seconds 1 milliseconds.
[2020-07-17 21:49:09,416] - [DEBUG] - 90.th resi

[2020-07-17 21:49:10,003] - [DEBUG] - __call__ took 0 seconds 2 milliseconds.
[2020-07-17 21:49:10,004] - [DEBUG] - 36.th residual: 0.1719151774639908
[2020-07-17 21:49:10,006] - [DEBUG] - __call__ took 0 seconds 2 milliseconds.
[2020-07-17 21:49:10,007] - [DEBUG] - 37.th residual: 0.1719151774639908
[2020-07-17 21:49:10,009] - [DEBUG] - __call__ took 0 seconds 2 milliseconds.
[2020-07-17 21:49:10,010] - [DEBUG] - 38.th residual: 0.1719151774639908
[2020-07-17 21:49:10,012] - [DEBUG] - __call__ took 0 seconds 2 milliseconds.
[2020-07-17 21:49:10,013] - [DEBUG] - 39.th residual: 0.1719151774639908
[2020-07-17 21:49:10,017] - [DEBUG] - __call__ took 0 seconds 3 milliseconds.
[2020-07-17 21:49:10,018] - [DEBUG] - 40.th residual: 0.1719151774639908
[2020-07-17 21:49:10,021] - [DEBUG] - __call__ took 0 seconds 2 milliseconds.
[2020-07-17 21:49:10,022] - [DEBUG] - 41.th residual: 0.1719151774639908
[2020-07-17 21:49:10,024] - [DEBUG] - __call__ took 0 seconds 2 milliseconds.
[2020-07-17 21:4

[2020-07-17 21:49:10,217] - [DEBUG] - __call__ took 0 seconds 3 milliseconds.
[2020-07-17 21:49:10,218] - [DEBUG] - 94.th residual: nan
[2020-07-17 21:49:10,221] - [DEBUG] - __call__ took 0 seconds 2 milliseconds.
[2020-07-17 21:49:10,221] - [DEBUG] - 95.th residual: nan
[2020-07-17 21:49:10,224] - [DEBUG] - __call__ took 0 seconds 2 milliseconds.
[2020-07-17 21:49:10,225] - [DEBUG] - 96.th residual: nan
[2020-07-17 21:49:10,228] - [DEBUG] - __call__ took 0 seconds 2 milliseconds.
[2020-07-17 21:49:10,228] - [DEBUG] - 97.th residual: nan
[2020-07-17 21:49:10,231] - [DEBUG] - __call__ took 0 seconds 2 milliseconds.
[2020-07-17 21:49:10,232] - [DEBUG] - 98.th residual: nan
[2020-07-17 21:49:10,236] - [DEBUG] - __call__ took 0 seconds 2 milliseconds.
[2020-07-17 21:49:10,236] - [DEBUG] - 99.th residual: nan
[2020-07-17 21:49:10,237] - [INFO] - HB-PCG iterations:   100
[2020-07-17 21:49:10,241] - [INFO] - Total time:     802 ms
[2020-07-17 21:49:10,241] - [INFO] - 


## Results

| Refinements | DoFs | Iterations | Computation time | Convergence rate |
| :---: | ------ | :---: | --- | :---: |
| 1 |     21 | ? | ? | ? |
| 2 |     65 | ? | ? | ? |
| 3 |    225 | ? | ? | ? |
| 4 |    833 | ? | ? | ? |
| 5 |   3201 | ? | ? | ? |
| 6 |  12545 | ? | ? | ? |
| 7 |  49665 | ? | ? | ? |
| 8 | 197633 | ? | ? | ? |
| 9 | 788481 | ? | ? | ? |